working/1426-extension-types/feature-specification.md

```diff
@@ -650,6 +650,13 @@ of `Never`.
 *In contrast to non-protected extension types, the extension type has no
 subtype relationship with its on-type.*
 
+It is a compile-time error if a protected extension type `E` is used to
+perform an explicit extension method invocation.
+
+*That is, `E(o).foo()` is an error when `E` is a protected extension
+type, also in the case where the static type of `o` matches the on-type of
+`E`. We can only access the members of `E` when the receiver has type `E`.*
+
```

Comment:
I would not say that as a blanket rule. 

The `E(o)` syntax matches a constructor invocation, so if `E` has an unnamed constructor taking `o` as argument, then `E(o).foo()` is precisely the way to call` `E.foo`.

The idea, as I see it, is instead that *non-protected* extensions get a default constructor of `factory E(OnType o) => o` (with or without the `factory`, since if it must always be there, it can be implicit). 
You cannot override or remove that constructor.
Then what was previously an "explicit extension invocation" now becomes an "explicit extension coercion" followed by a normal member call on the extension type.

Protected extension types do not get such a default constructor if you declare any other constructor.

Reply:
More or less what @lrhn said, though I'm not sure I'm sold on the various iterations of what gets a default when.

Reply:
I think we _would_ want to make an explicit extension method invocation on an instance of the on-type an error: That instance hasn't been vetted.

However, we're using the same syntax for an explicit extension method invocation and for an extension type constructor invocation as a receiver in a member invocation. So we need to be able to make that distinction clear.

We also need to be able to specify an explicit extension method invocation in order to specify the semantics of a member invocation where the receiver type is an extension type.

I've introduced a new primitive `invokeExtensionMethod`, which is intended to be a specification device (not actual Dart syntax), which is always denoting an extension method invocation. That solves several problems, I think.

So let's continue this discussion based on the new text.